src/types/Utf8Utf16Convert.cpp

```diff
@@ -383,94 +252,31 @@ void u16state::reset() noexcept
 // Arguments:
 // - in - wstring_view of the UTF-16 string to be converted
 // - out - reference to the resulting UTF-8 string
-// - discardInvalids - if `true` invalid characters are discarded instead of replaced with U+FFFD, default is `false`
 // Return Value:
-// - S_OK          - the conversion succeded without any change of the represented code points
-// - S_FALSE       - the incoming string contained at least one invalid character
+// - S_OK          - the conversion succeded
 // - E_OUTOFMEMORY - the function failed to allocate memory for the resulting string
-// - E_ABORT       - the resulting string length would exceed the max_size and thus, the conversion was aborted before the conversion has been completed
+// - E_ABORT       - the resulting string length would exceed the upper boundary of an int and thus, the conversion was aborted before the conversion has been completed
 // - E_UNEXPECTED  - an unexpected error occurred
-[[nodiscard]] HRESULT u16u8(const std::wstring_view in, std::string& out, bool discardInvalids) noexcept
+[[nodiscard]] HRESULT til::u16u8(const std::wstring_view in, std::string& out) noexcept
 {
-    constexpr const uint32_t unicodeReplacementChar{ 0xFFFDu };
-
     try
     {
-        HRESULT hRes{ S_OK };
         out.clear();
 
         if (in.empty())
         {
-            return hRes;
-        }
-
-        size_t lengthHint{};
-        if (FAILED(SizeTMult(in.length(), gsl::narrow_cast<size_t>(3u), &lengthHint)))
-        {
-            lengthHint = std::max(out.capacity(), in.length());
+            return S_OK;
         }
 
-        out.reserve(lengthHint); // avoid any further re-allocations and copying
+        int lengthIn{};
+        int lengthRequired{};
+        RETURN_HR_IF(E_ABORT, FAILED(SizeTToInt(in.length(), &lengthIn)) || FAILED(IntMult(lengthIn, 3, &lengthRequired)));
 
-        const auto end16{ in.cend() };
-        for (auto it16{ in.cbegin() }; it16 < end16;)
-        {
-            uint32_t codePoint{ unicodeReplacementChar }; // default
+        out.resize(gsl::narrow_cast<size_t>(lengthRequired)); // avoid to call WideCharToMultiByte twice only to get the required size
+        const int lengthOut = WideCharToMultiByte(65001u, 0ul, in.data(), lengthIn, out.data(), lengthRequired, nullptr, nullptr);
+        out.resize(gsl::narrow_cast<size_t>(lengthOut));

```

Comment:
Why is narrow_cast always safe here?

Reply:
According to the manpage of WC2MB the return value is either the number of bytes written or 0. So, there is no negative value returned. That the worst case size fits into an `int` has been already evaluated for the `lengthRequired` variable, the return value can't be greater than that. `int` has _at least_ the same size as `size_t` and thus, a positive `int` will never lead to change the value if casted to `size_t`.